File: swot.jl
### Run SAD algorithm with SWOT data

using Sad
using ArgParse
using DelimitedFiles
using Distributions
using JSON
using LinearAlgebra
using NCDatasets
using PyCall

const FILL = -999999999999

"""
    get_reach_files(indir, reachjson)

Get reach file names and download SoS file.

"""
function get_reach_files(indir, tmpdir, reachjson, index, sosbucket)
    data = Ref{Dict{String, Any}}
    open(joinpath(indir, reachjson)) do jf
        data = read(jf, String)
    end

    reachlist = JSON.parse(data)[index]
    if !isempty(sosbucket)
        sosfile = joinpath(tmpdir, reachlist["sos"])
        pushfirst!(pyimport("sys")."path", "./sos_read")   # Load sos_read script
        downloadsos = pyimport("sos_read")["download_sos"]
        downloadsos(sosbucket, sosfile)
        reachlist["reach_id"], joinpath(indir, "swot", reachlist["swot"]), sosfile, joinpath(indir, "sword", reachlist["sword"])
    else
        reachlist["reach_id"], joinpath(indir, "swot", reachlist["swot"]), joinpath(indir, "sos", reachlist["sos"]), joinpath(indir, "sword", reachlist["sword"])
    end
end

"""
    read_swot_obs(ncfile, nids)

Load SWOT observations.

# Arguments

- `ncfile`: NetCDF file with SWOT observations
- `nids`: sorted node IDs from downstream to upstream

"""
function read_swot_obs(ncfile::String, nids::Vector{Int})
    Dataset(ncfile) do ds
        nodes = NCDatasets.group(ds, "node")
        reaches = NCDatasets.group(ds, "reach")
        S = permutedims(nodes["slope2"][:])
        H = permutedims(nodes["wse"][:])
        W = permutedims(nodes["width"][:])
        dA = reaches["d_x_area"][:]
        dA = convert(Vector{Sad.FloatM}, dA)
        Hr = convert(Vector{Sad.FloatM}, reaches["wse"][:])
        Wr = convert(Vector{Sad.FloatM}, reaches["width"][:])
        Sr = convert(Vector{Sad.FloatM}, reaches["slope2"][:])
        W[.!ismissing.(H) .&& isnan.(H)] .= missing
        S[.!ismissing.(H) .&& isnan.(H)] .= missing
        H[.!ismissing.(H) .&& isnan.(H)] .= missing
        nid = nodes["node_id"][:]
        dmap = Dict(nid[k] => k for k=1:length(nid))
        i = [dmap[k] for k in nids]
        H[i, :], W[i, :], S[i, :], dA, Hr, Wr, Sr
    end
end

"""
    river_info(id, swordfile)

Retrieve information about river reach cross sections.

# Arguments

- `id`: reach ID
= `swordfile`: SWORD NetCDF file

"""
function river_info(id::Int, swordfile::String)
    Dataset(swordfile) do fd
        g = NCDatasets.group(fd, "nodes")
        i = findall(g["reach_id"][:] .== id)
        nid = g["node_id"][i]
        x = g["dist_out"][i]
        k = findall(.!ismissing.(nid))
        x = x[k]
        nid = nid[k]
        # subtract the minimum `dist_out` and then sort from downstream to upstream
        x = x .- minimum(x)
        j = sortperm(x)
        close(fd)
        convert(Vector{Int}, nid[j]), x[j]
    end
end

"""
    write_output(reachid, valid, outdir, A0, n, Qa, Qu)

Write SAD output to NetCDF.

"""
function write_output(reachid, valid, outdir, A0, n, Qa, Qu, W)
    outfile = joinpath(outdir, "$(reachid)_sad.nc")
    out = Dataset(outfile, "c")
    out.attrib["valid"] = valid   # FIXME Determine what is considered valid in the context of a SAD run
    defDim(out, "nx", size(W, 1))
    defDim(out, "nt", size(W, 2))
    ridv = defVar(out, "reach_id", Int64, (), fillvalue = FILL)
    ridv[:] = reachid
    A0v = defVar(out, "A0", Float64, (), fillvalue = FILL)
    A0v[:] = A0
    nv = defVar(out, "n", Float64, (), fillvalue = FILL)
    nv[:] = n
    Qav = defVar(out, "Qa", Float64, ("nt",), fillvalue = FILL)
    Qav[:] = replace!(Qa, NaN=>FILL)
    Quv = defVar(out, "Q_u", Float64, ("nt",), fillvalue = FILL)
    Quv[:] = Qu
    close(out)
end

"""
    parse_commandline(r)

Parse command line for arguments.

"""
function parse_commandline()
    s = ArgParseSettings()
    @add_arg_table s begin
        "--index", "-i"
            help = "Index of reach to run on"
            arg_type = Int
            default = 0
        "--reachfile", "-r"
            help = "Name of reaches JSON file"
            arg_type = String
            default = "reaches.json"
        "--bucketkey", "-b"
            help = "Bucket and key prefix to download SoS from"
            arg_type = String
            default = ""
    end

    return parse_args(s)
end

"""
    main()

Main driver routine.

"""
function main()
    indir = joinpath("/mnt", "data", "input")
    outdir = joinpath("/mnt", "data", "output")
    tmpdir = joinpath("/tmp")

    parsed_args = parse_commandline()
    index = parsed_args["index"] + 1
    reachfile = parsed_args["reachfile"]
    bucketkey = parsed_args["bucketkey"]
    println("Index: $(index)")
    println("Reach File: $(reachfile)")
    println("Bucket Key: $(bucketkey)")

    reachid, swotfile, sosfile, swordfile = get_reach_files(indir, tmpdir, reachfile, index, bucketkey)
    println("Reach ID: $(reachid)")
    println("SWOT: $(swotfile)")
    println("SOS: $(sosfile)")
    println("SWORD: $(swordfile)")

    nids, x = river_info(reachid, swordfile)
    H, W, S, dA, Hr, Wr, Sr = read_swot_obs(swotfile, nids)

    try
        x, H, W, S = Sad.drop_unobserved(x, H, W, S)
    catch e
        if e isa MethodError
            println("Error loading swot observation")
            end
        end
    A0 = missing
    n = missing
    Qa = Array{Missing}(missing, 1, size(W, 2))
    Qu = Array{Missing}(missing, 1, size(W, 2))
    if all(ismissing, H) || all(ismissing, W) || all(ismissing, S)
        println("$(reachid): INVALID")
        write_output(reachid, 0, outdir, A0, n, Qa, Qu, W)
    else
        Hmin = minimum(skipmissing(H[1, :]))
        Qp, np, rp, zp = Sad.priors(sosfile, Hmin, reachid)
        if ismissing(Qp)
            println("$(reachid): INVALID, missing mean discharge")
            write_output(reachid, 0, outdir, A0, n, Qa, Qu, W)
        else
            try
                nens = 100 # default ensemble size
                nsamples = 1000 # default sampling size
                Qa, Qu, A0, n = Sad.estimate(x, H, W, S, dA, Qp, np, rp, zp, nens, nsamples, Hr, Wr, Sr)
                println("$(reachid): VALID")
                write_output(reachid, 1, outdir, A0, n, Qa, Qu, W)
            catch
                println("$(reachid): INVALID")
                write_output(reachid, 0, outdir, A0, n, Qa, Qu, W)
            end
        end
    end
end

main()

File: swot.py
"""
swot.py

Main driver script for the SAD algorithm with SWOT data.
Translated from swot.jl.
"""

import argparse
import json
import os
import sys
import traceback

import numpy as np
import netCDF4 as nc

import jax
jax.config.update("jax_enable_x64", True)

from preprocess import preprocess
from priors import priors_from_sos, priors_from_qwbm, River
from infer import infer
from utils import strip_mask

FILL = -999999999999


# ---------------------------------------------------------------------------
# File discovery
# ---------------------------------------------------------------------------

def get_reach_files(indir, tmpdir, reachjson, index, sosbucket):
    """
    Get reach file names and optionally download SoS file from S3.
    """
    with open(os.path.join(indir, reachjson)) as f:
        reachlist = json.load(f)[index]

    if sosbucket:
        sosfile = os.path.join(tmpdir, reachlist["sos"])
        sys.path.insert(0, "/app/sos_read")
        from sos_read import download_sos
        download_sos(sosbucket, sosfile)
    else:
        sosfile = os.path.join(indir, "sos", reachlist["sos"])

    return (
        reachlist["reach_id"],
        os.path.join(indir, "swot",  reachlist["swot"]),
        sosfile,
        os.path.join(indir, "sword", reachlist["sword"]),
    )


# ---------------------------------------------------------------------------
# SWORD reader
# ---------------------------------------------------------------------------

def river_info(reach_id, swordfile):
    """
    Retrieve node IDs and chainage from SWORD NetCDF file.
    Returns (nids, x) sorted downstream -> upstream.
    """
    with nc.Dataset(swordfile) as fd:
        g       = fd.groups["nodes"]
        mask    = g.variables["reach_id"][:] == reach_id
        nid     = g.variables["node_id"][mask]
        x       = g.variables["dist_out"][mask]

        # Drop missing node IDs
        valid   = ~np.ma.getmaskarray(nid)
        nid     = strip_mask(nid[valid]).astype(int)
        x       = strip_mask(x[valid])

        # Reset chainage to start from 0 and sort downstream -> upstream
        x      -= x.min()
        order   = np.argsort(x)
        return nid[order], x[order]


# ---------------------------------------------------------------------------
# SWOT reader
# ---------------------------------------------------------------------------

def read_swot_obs(ncfile, nids):
    """
    Load SWOT node-level observations, ordered by nids (downstream -> upstream).

    Returns H, W, S as (nobs x nt) float64 arrays with NaN sentinels,
    and time_str as a list of strings.
    """
    with nc.Dataset(ncfile) as ds:
        nodes   = ds.groups["node"]
        reaches = ds.groups["reach"]

        # Node-level arrays: stored as (nt x nnodes) -> transpose to (nnodes x nt)
        S_raw   = strip_mask(nodes.variables["slope2"][:]).T
        H_raw   = strip_mask(nodes.variables["wse"][:]).T
        W_raw   = strip_mask(nodes.variables["width"][:]).T

        # Map node IDs to row indices
        node_ids = strip_mask(nodes.variables["node_id"][:]).astype(int)
        id_to_idx = {nid: k for k, nid in enumerate(node_ids)}
        rows = [id_to_idx[n] for n in nids if n in id_to_idx]

        H = H_raw[rows]
        W = W_raw[rows]
        S = S_raw[rows]

        # NaN-out W and S where H is NaN (consistency)
        nan_h       = np.isnan(H)
        W[nan_h]    = np.nan
        S[nan_h]    = np.nan

        # Reach-level time strings
        time_var    = reaches.variables["time_str"]
        time_raw    = np.array(time_var[:])
        if time_raw.ndim == 2:
            time_str = ["".join(row).strip() for row in time_raw]
        else:
            time_str = [str(t).strip() for t in time_raw]

        return H, W, S, time_str


# ---------------------------------------------------------------------------
# Output writer
# ---------------------------------------------------------------------------

def write_output(reachid, valid, outdir, A0, n, Qa, Qu, W, time_str):
    """Write SAD output to NetCDF."""
    outfile = os.path.join(outdir, f"{reachid}_sad.nc")
    nt      = W.shape[1] if W.ndim == 2 else len(time_str)
    nx      = W.shape[0] if W.ndim == 2 else 1

    with nc.Dataset(outfile, "w") as out:
        out.valid = int(valid)

        out.createDimension("nx", nx)
        out.createDimension("nt", nt)

        rid_v           = out.createVariable("reach_id", "i8", (), fill_value=FILL)
        rid_v[:]        = reachid

        A0_v            = out.createVariable("A0",  "f8", (), fill_value=FILL)
        A0_v[:]         = float(A0) if A0 is not None else FILL

        n_v             = out.createVariable("n",   "f8", (), fill_value=FILL)
        n_v[:]          = float(n)  if n  is not None else FILL

        Qa_v            = out.createVariable("Qa",  "f8", ("nt",), fill_value=FILL)
        Qa_arr          = np.asarray(Qa, dtype=np.float64)
        Qa_arr[np.isnan(Qa_arr)] = FILL
        Qa_v[:]         = Qa_arr

        Qu_v            = out.createVariable("Q_u", "f8", ("nt",), fill_value=FILL)
        Qu_arr          = np.asarray(Qu, dtype=np.float64)
        Qu_arr[np.isnan(Qu_arr)] = FILL
        Qu_v[:]         = Qu_arr

        ts_v            = out.createVariable("time_str", str, ("nt",))
        ts_v[:]         = np.array(time_str[:nt], dtype=object)


# ---------------------------------------------------------------------------
# Posterior uncertainty
# ---------------------------------------------------------------------------

def posterior_uncertainty(res, nt):
    """
    Compute per-timestep posterior Q uncertainty (std dev) from the
    ensemble. Returns NaN where the timestep was skipped.
    """
    Qu = np.full(nt, np.nan)
    for t, A in enumerate(res["A_post"]):
        if A is not None:
            Qu[t] = float(A[0].std())
    return Qu


# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------

def parse_args():
    p = argparse.ArgumentParser(description="SAD algorithm with SWOT data")
    p.add_argument("--index", "-i", type=int, default=0,
                   help="Index of reach to run on")
    p.add_argument("--reachfile", "-r", type=str, default="reaches.json",
                   help="Name of reaches JSON file")
    p.add_argument("--bucketkey", "-b", type=str, default="",
                   help="Bucket and key prefix to download SoS from S3")
    return p.parse_args()


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    indir  = os.path.join("/mnt", "data", "input")
    outdir = os.path.join("/mnt", "data", "output")
    tmpdir = "/tmp"

    args = parse_args()

    # AWS Batch array job index handling
    if args.index == -256:
        try:
            index = int(os.environ["AWS_BATCH_JOB_ARRAY_INDEX"])
        except KeyError:
            index = 0
    else:
        index = args.index

    reachfile = args.reachfile
    bucketkey = args.bucketkey
    print(f"Index: {index}")
    print(f"Reach File: {reachfile}")
    print(f"Bucket Key: {bucketkey}")

    reachid, swotfile, sosfile, swordfile = get_reach_files(
        indir, tmpdir, reachfile, index, bucketkey,
    )
    print(f"Reach ID: {reachid}")
    print(f"SWOT:     {swotfile}")
    print(f"SOS:      {sosfile}")
    print(f"SWORD:    {swordfile}")

    nids, x = river_info(reachid, swordfile)
    H, W, S, time_str = read_swot_obs(swotfile, nids)
    nt = H.shape[1]

    # Default invalid outputs
    A0 = None
    n  = None
    Qa = np.full(nt, np.nan)
    Qu = np.full(nt, np.nan)

    if np.all(np.isnan(H)) or np.all(np.isnan(W)) or np.all(np.isnan(S)):
        print(f"{reachid}: INVALID")
        write_output(reachid, 0, outdir, A0, n, Qa, Qu, W, time_str)
        return

    try:
        reach  = preprocess(x, H, W, S)
        hmin   = reach.hmin
        priors = priors_from_sos(sosfile, hmin, reachid)

        if priors is None:
            print(f"{reachid}: INVALID, missing mean discharge in SoS")
            write_output(reachid, 0, outdir, A0, n, Qa, Qu, W, time_str)
            return

        res = infer(priors, reach)

        Qa = res["Q_post"]
        Qu = posterior_uncertainty(res, nt)

        # Posterior reach parameter means from the reach ensemble
        reach_ens = res["reach_ensemble"]
        n  = float(reach_ens[0].mean())   # Manning n
        r  = float(reach_ens[1].mean())   # Dingman r
        z0 = float(reach_ens[2].mean())   # bed elevation
        # A0 = bankfull cross-sectional area at downstream node
        # Using Dingman formula at bankfull depth
        hbf_ds = float(reach.hbf(reach.x[0]))
        Yb_ds  = max(hbf_ds - z0, 0.01)
        Wb_ds  = float(reach.wbf(reach.x[0]))
        Ym_ds  = (r + 1) / r * Yb_ds
        A0     = Wb_ds * (Ym_ds / Yb_ds) ** (1.0 / r) * Yb_ds

        print(f"{reachid}: VALID")
        write_output(reachid, 1, outdir, A0, n, Qa, Qu, W, time_str)

    except Exception:
        traceback.print_exc()
        print(f"{reachid}: INVALID")
        write_output(reachid, 0, outdir, A0, n, Qa, Qu, W, time_str)


if __name__ == "__main__":
    main()
